PR_Review_V1.0

KiranQode opened the pull request.

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,107 @@
+parent.geometry('300x50')
+frame = tk.Frame(parent)
+frame.pack()
+fileButton = tk.Button(frame, text="Select CANData file(.txt)", fg="Green", command=lambda: CAN_Decoder())
```
You could use `command=CAN_Decoder`. Because you put () after `CAN_Decoder`, insted of passing the function that needs to be called on event, you called the function and passed its result.

KiranQode replied:
Updated. Please refer V1.1

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,107 @@
+    lineNum = 0
+    oldDataFile = filedialog.askopenfile(initialdir="/")
+    listedOldData = []
+    for dataLine in oldDataFile:
```
try using Regex for extracting information from a text file. 
Regex is de facto standard for parsing text and is a must know library in most of programming languages.
This is a good start: https://www.w3schools.com/python/python_regex.asp
you can practice here: https://regex101.com/
If you need help with the complete regex for extracting the information, just let me know

KiranQode replied:
Thanks for the resources shared. I have gone through the resources but I am not sure how easy we can use it for our case. Please suggest.

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,107 @@
+
+    newData = []
+    lineCount = len(listedOldData)
+    for dataLineIndex in range(0, lineCount):
```
Why don't you iterate over `listedOldData` directly ? 
`for dataLine in listedOldData`

KiranQode replied:
lineCount is re-used while parsing data  for consecutive frames

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,107 @@
+        listedOldData.append(listedOldDataLine)
+        lineNum = lineNum + 1
+
+    newData = []
```
A good idea will be to separete the "input data parsing" from "TP frame reassembly" into 2 separate functions.

KiranQode replied:
Implemented. Please refer V1.1

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,107 @@
+    for dataLineIndex in range(0, lineCount):
+        dataLine = listedOldData[dataLineIndex]
+        refinedDataLine = []
+        frameType = int(dataLine[3])
```
it's better to transform all data to int before start the ISO-TP reassembly

KiranQode replied:
All payload data are not convertible to int  for e.g. 1C, 6C, etc.  Please suggest

dmgciubotaru commented:
As a general note, try to transform the CAN frame from string to a more handy type.
For this frame "  can0  7DF   [8]  02 01 00 00 00 00 00 00"
You can use tuple : (0, 2015, [2,1,0,0,0,0,0,0])
Or a dict {'bid':0, 'mid': 2015, 'data':[2,1,0,0,0,0,0,0]}
this will make the data manipulation much more asier

KiranQode commented:
> As a general note, try to transform the CAN frame from string to a more handy type.
> For this frame " can0 7DF [8] 02 01 00 00 00 00 00 00"
> You can use tuple : (0, 2015, [2,1,0,0,0,0,0,0])
> Or a dict {'bid':0, 'mid': 2015, 'data':[2,1,0,0,0,0,0,0]}
> this will make the data manipulation much more asier

Implemented: Added string operations to remove whitespaces and converted elements into array before creating dictionary. Please comment.

KiranQode commented:
[CAN_Decoder_V1.1.py.zip](https://github.com/KiranQode/CANData_DeCoder_Python/files/7834894/CAN_Decoder_V1.1.py.zip)

KiranQode commented:
Created new pull request for review

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,119 @@
+    typicalFrameForm = r"can(\d)\s+([(\w)]{1,3})\s+\[(\d+)\]\s+([\w]{2}(?:\s[\w]{2})*)"
+    frameSize = 0
+    for dataFrame in re.findall(typicalFrameForm, open(dataFile).read()):
+        diagnosticID = int(dataFrame[1], 16)
```
this is rather the "messageID". The diagnostic comes into play after the messages are reassemled (but this is a minor thing)

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,119 @@
+    for dataFrame in re.findall(typicalFrameForm, open(dataFile).read()):
+        diagnosticID = int(dataFrame[1], 16)
+        data = list(map(lambda x: int(x, 16), dataFrame[3].split(" ")))
+        frameType = data[0]
```
the data[0] is of form "XY" (hex) where "X" is the frame type and "Y"'s meaning is dependent on "X".
To extract the "X" you can do `frameType = data[0] // 16`

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,119 @@
+        diagnosticID = int(dataFrame[1], 16)
+        data = list(map(lambda x: int(x, 16), dataFrame[3].split(" ")))
+        frameType = data[0]
+        if frameType <= 6:
```
This is right for this example, but for can messages larger than 8 bytes, this will not work.
The check should be `data[0] // 16 == 0`

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,119 @@
+        frameType = data[0]
+        if frameType <= 6:
+            frameSize = frameType  # Single frame
+        elif frameType == 16:  # 10 for hex notation
```
based on ISO-TP documentation if the frame starts with `1A BC`  the total frame size will be `ABC`. In these examples `A` just happended to be always 0. I will let you figure out a way to extract `A` from first byte and append it to `BC`

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,119 @@
+    inputData_dict = {}
+    newLineCount = 0
+    cFrameCount = 0
+    for dataLineIndex in range(0, lineCount):
```
you can use `enumerate` here: `for dataLine, dataLineIndex in enumerate(formattedRawData):`

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,119 @@
+        if frameType <= 6:      # Single frame
+            sFrameIndex = 1     # Single frame index
+            payloadData = []
+            for i in range(sFrameIndex, sFrameIndex + frameSize):
```
list slicing can come in handy here: `dataLine["PayloadData"][sFrameIndex : sFrameIndex + frameSize + 1]` will extract from `PayloadData` the portion you need. 
https://www.geeksforgeeks.org/python-list-slicing/

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,119 @@
+            eleCount = 0
+            fFrameIndex = 2     # First frame index
+            payloadData = []
+            for j in range(fFrameIndex, len(dataLine["PayloadData"])):
```
use list slicing. If you don;t specify the upper bound, it will take the entire vector: `dataLine["PayloadData"][ fFrameIndex : ]`

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,119 @@
+                    if seqNum != 16:   # 10 for hex notation
+                        seqCount += 1
+                        if seqNum == startSeqNum + seqCount:
+                            for j in range(cFrameIndex, len(dataLine_CF["PayloadData"])):
```
use list slicing

dmgciubotaru commented on CAN_Decoder_SourceCode_TestRel_21Dec21.py:
```diff
@@ -0,0 +1,119 @@
+                diagnosticID_CF = dataLine_CF["DiagnosticID"]
+                if diagnosticId == diagnosticID_CF:
+                    seqNum = dataLine_CF["PayloadData"][0]
+                    if seqNum != 16:   # 10 for hex notation
```
you can write hex numbers : `if seqNum != 0x10`

dmgciubotaru reviewed (changes requested):
Another general note, try to use shorter names for variables. You will end up losing a lot of time writing em.
